File: src/tree/js/BinaryTreePruning.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */

var pruneTree = function(root) {
    return purgeZeros(root)!==0?root:null;
};

let purgeZeros = node =>{
    if(node === null){
        return 0;
    }

    let left = purgeZeros(node.left);
    if (left === 0){
        node.left = null;
    }

    let right = purgeZeros(node.right);
    if (right === 0){
        node.right = null;
    }

    let total = left + right + node.val;

    return total;
}

File: src/tree/BinaryTreePruning.java
package tree;

public class BinaryTreePruning {
    public static void main(String[] args) {

    }

    public TreeNode pruneTree(TreeNode root) {
        return purgeZeros(root)!=0?root:null;
    }

    private int purgeZeros(TreeNode root) {
        if(root==null)
            return 0;

        int left = purgeZeros(root.left);
        if(left == 0){
            root.left=null;
        }
        int right = purgeZeros(root.right);
        if(right == 0){
            root.right=null;
        }
        int total = left + right + root.val;
        return total;
    }
}
